### PR TITLE
Ensure FileStream.Position is correct after a failed|cancelled WriteAsync attempt

### PR DESCRIPTION
**Pull request by adamsitnik**
@stephentoub I've realised that while writes can't be incomplete, they might fail with an exception and in such cases we should update the position as well.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs [hunk 1]
```diff
@@ -116,8 +118,16 @@ private void ExecuteInternal()
                 }
                 finally
                 {
+                    if (_strategy is not null && exception is not null)
+                    {
+                        Debug.Assert(_operation == Operation.Write);
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs:123]: we do it only for writes as `OSFileStreamStrategy.ReadAsync` is not using this type:

https://github.com/dotnet/runtime/blob/79c4144f174754bf4fa18827a292549214f86581/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/OSFileStreamStrategy.cs#L310-L322

File: src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs [hunk 2]
```diff
@@ -248,7 +248,7 @@ internal static ValueTask<int> ReadAtOffsetAsync(SafeFileHandle handle, Memory<b
             handle.EnsureThreadPoolBindingInitialized();
 
             SafeFileHandle.OverlappedValueTaskSource vts = handle.GetOverlappedValueTaskSource();
-            int errorCode = 0;
+            int errorCode = Interop.Errors.ERROR_SUCCESS;
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs:251]: addresses https://github.com/dotnet/runtime/pull/56531#discussion_r680398237

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs [hunk 3]
```diff
@@ -43,28 +43,17 @@ public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationT
...
             long writeOffset = CanSeek ? Interlocked.Add(ref _filePosition, buffer.Length) - buffer.Length : -1;
 
-            (SafeFileHandle.OverlappedValueTaskSource? vts, int errorCode) = RandomAccess.QueueAsyncWriteFile(_fileHandle, buffer, writeOffset, cancellationToken);
+            (SafeFileHandle.OverlappedValueTaskSource? vts, int errorCode) = RandomAccess.QueueAsyncWriteFile(_fileHandle, buffer, writeOffset, cancellationToken, this);
             return vts != null
                 ? new ValueTask(vts, vts.Version)
-                : (errorCode == 0) ? ValueTask.CompletedTask : ValueTask.FromException(HandleIOError(writeOffset, errorCode));
-        }
-
-        private Exception HandleIOError(long positionBefore, int errorCode)
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs:59]: the `HandleIOError` method is not needed anymore as all sync (immediately known) errors are handled at `RandomAccess.QueueAsyncReadFile` and `RandomAccess.QueueAsyncWriteFile` while async errors are handled by `OverlappedValueTaskSource.Complete`

**Comment by stephentoub** [comment]
> I've realised that while writes can't be incomplete, they might fail with an exception and in such cases we should update the position as well.

Just to clarify, this never used to be done in previous .NET releases, right?

File: src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs:123]: > we do it only for writes as OSFileStreamStrategy.ReadAsync is not using this type

Per my previous comment... should it?

**Comment by stephentoub** [comment]
> I think it was since .NET Core 3.1 (as least to some degree):

Isn't your example only showing it actually happening when the token is canceled in advance, which means the operation never actually starts?

Or, if I'm misunderstanding the example, can you point to what code was updating the position after a failure?

File: src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs [hunk 4]
```diff
@@ -32,6 +33,7 @@ internal sealed class ThreadPoolValueTaskSource : IThreadPoolWorkItem, IValueTas
             private ManualResetValueTaskSourceCore<long> _source;
             private Operation _operation = Operation.None;
             private ExecutionContext? _context;
+            private OSFileStreamStrategy? _strategy;
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs:36]: that was a great advice, #done

**Comment by adamsitnik** [comment]
@stephentoub you are right, so far we were not updating the position after failure

FWIW I've checked why in my sample we fail to cancel to `WriteFile` request. The code calls `CancelIoEx` using proper arguments:

https://github.com/dotnet/runtime/blob/b3ab2eb4db830f6c520340f80651b806fd9dfce8/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.OverlappedValueTaskSource.Windows.cs#L124

But it fails with `ERROR_NOT_FOUND `, which we have been always ignoring:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/Net5CompatFileStreamStrategy.CompletionSource.Windows.cs#L205-L206

**Comment by adamsitnik** [comment]
Closing an re-opening to try to trigger the CI as the last x64 debug Windows [build](https://dnceng.visualstudio.com/public/_build/results?buildId=1272865&view=results) failed with:

> Failed to download package 'System.Reflection.Metadata.1.8.0'

[full log](https://dnceng.visualstudio.com/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/1272865/logs/440)

**Review by stephentoub — APPROVED** [review]

**Pull request merged**